namer/consul/src/main/scala/io/buoyant/namer/consul/SvcAddr.scala

```diff
@@ -0,0 +1,104 @@
+package io.buoyant.namer.consul
+
+import com.twitter.finagle._
+import com.twitter.util._
+import io.buoyant.consul.v1
+import io.buoyant.namer.Metadata
+import scala.util.control.NoStackTrace
+
+private[consul] case class SvcKey(name: String, tag: Option[String]) {
+  override def toString = tag match {
+    case Some(t) => s"$name:$t"
+    case None => name
+  }
+}
+
+private[consul] object SvcAddr {
+
+  /**
+   * Runs a long-polling loop on a service object to obtain the set of
+   * Addresses.  This evaluates lazily so that only activity observed
+   */
+  def apply(
+    consulApi: v1.ConsulApi,
+    datacenter: String,
+    key: SvcKey,
+    domain: Option[String]
+  ): Var[Addr] = {
+
+    def getAddresses(index: Option[String]): Future[v1.Indexed[Set[Address]]] =
+      consulApi.serviceNodes(
+        key.name,
+        datacenter = Some(datacenter),
+        tag = key.tag,
+        blockingIndex = index,
+        retry = true
+      ).map(indexedToAddresses)
+
+    // Start by fetching the service immediately, and then long-poll
+    // the service for updates.
+    Var.async[Addr](Addr.Pending) { state =>
+      @volatile var stopped: Boolean = false
+
+      def loop(index0: Option[String]): Future[Unit] =
+        if (stopped) Future.Unit
```

Comment:
As discussed in #625, it's necessary to use guards to ensure that work stops eventually.